File: web/src/customVoice.ts
import { api } from "./api";
import { modelAuthHeaders } from "./llmModels";
import type { LlmProviderId } from "./llmProviders";

const MAX_EXTRA = 2048;
const SYSTEM =
  "You write a Klar NLU voice block only. " +
  "If a seed character is given, keep that identity. " +
  "Sliders 0-10 only refine delivery (warmth, humor, sarcasm, formality, verbosity, energy). " +
  "They must not replace or ignore the seed. " +
  "Output the voice description and a few short example rewrites (source → spoken). " +
  "Keep language lock and safety: no Home Assistant tools, no device control, " +
  "digits stay digits, no new facts. Do not repeat the safety rules. " +
  "Do not mention this instruction. No markdown fences. No title.";

export type CustomVoiceIn = {
  language: string;
  address: string;
  name?: string;
  voice_name?: string;
  seed?: string;
  warmth: number;
  humor: number;
  sarcasm: number;
  formality: number;
  verbosity: number;
  energy: number;
  taboo?: string;
};

export type LlmChatTarget = {
  baseUrl: string;
  model: string;
  apiKey: string;
  provider: LlmProviderId;
};

export async function writeCustomVoice(body: CustomVoiceIn, llm?: LlmChatTarget): Promise<string> {
  try {
    const out = await api.customVoice(body);
    if (out.prompt.trim()) return clipVoice(out.prompt);
  } catch {
    if (!llm?.baseUrl.trim() || !llm.model.trim()) throw new Error("voice");
  }
  if (!llm?.baseUrl.trim() || !llm.model.trim()) throw new Error("voice");
  return clipVoice(await chatDirect(llm, body));
}

function languageLock(pack: string): string {
  const tag = pack.trim().toLowerCase();
  if (tag === "de" || tag.startsWith("de-")) {
    return "Antworte nur auf Deutsch. Übersetze nicht ins Englische oder in eine andere Sprache.";
  }
  if (tag === "en" || tag.startsWith("en-")) {
    return "Answer only in English. Do not translate into German or any other language.";
  }
  return `Answer only in the Klar NLU pack ${pack}. Do not translate into German, English, or any other language.`;
}

function interviewLine(body: CustomVoiceIn): string {
  const address =
    body.address === "name" ? `address by first name (${(body.name || "").trim()})` : body.address;
  const voiceName = (body.voice_name || "").trim() || "custom";
  let line =
    `Voice name: ${voiceName}.\nLanguage pack: ${body.language}.\nAddress the operator: ${address}.\n` +
    `Traits 0-10 (delivery only): warmth=${body.warmth}, humor=${body.humor}, sarcasm=${body.sarcasm}, ` +
    `formality=${body.formality}, verbosity=${body.verbosity}, energy=${body.energy}.`;
  const seed = (body.seed || "").trim();
  if (seed) line += `\nSeed character (keep this identity; sliders only refine delivery):\n${seed}`;
  const taboo = (body.taboo || "").trim();
  if (taboo) line += `\nDo not say: ${taboo}`;
  return line;
}

function chatUrls(baseUrl: string): string[] {
  const base = baseUrl.trim().replace(/\/+$/, "");
  const urls = [`${base}/chat/completions`];
  if (base.endsWith("/api/v1")) {
    urls.push(`${base.slice(0, -"/api/v1".length)}/v1/chat/completions`);
  } else if (base.endsWith("/v1")) {
    urls.push(`${base.slice(0, -"/v1".length)}/api/v1/chat/completions`);
  }
  return [...new Set(urls)];
}

function noThinking(provider: LlmProviderId): Record<string, unknown> {
  switch (provider) {
    case "openai":
      return {};
    case "anthropic":
      return {};
    case "google":
      return {
        reasoning_effort: "none",
        extra_body: { google: { thinking_config: { thinking_budget: 0 } } },
      };
    default:
      return { chat_template_kwargs: { enable_thinking: false } };
  }
}

function completionText(value: unknown): string {
  if (!value || typeof value !== "object") return "";
  const choices = (value as { choices?: unknown }).choices;
  if (!Array.isArray(choices) || !choices[0] || typeof choices[0] !== "object") return "";
  const message = (choices[0] as { message?: { content?: unknown } }).message;
  const content = message?.content;
  return typeof content === "string" ? content : "";
}

async function chatDirect(llm: LlmChatTarget, body: CustomVoiceIn): Promise<string> {
  const key = llm.apiKey.trim() || (llm.provider === "lemonade" ? "lemonade" : "");
  const headers: Record<string, string> = {
    ...(modelAuthHeaders(llm.provider, key) as Record<string, string>),
    "Content-Type": "application/json",
  };
  const payload = {
    model: llm.model.trim(),
    messages: [
      { role: "system", content: `${SYSTEM}\n\n${languageLock(body.language)}` },
      { role: "user", content: interviewLine(body) },
    ],
    stream: false,
    temperature: 0.4,
    max_tokens: 512,
    ...noThinking(llm.provider),
  };
  for (const url of chatUrls(llm.baseUrl)) {
    try {
      const response = await fetch(url, { method: "POST", headers, body: JSON.stringify(payload) });
      if (!response.ok) continue;
      const text = completionText(await response.json());
      if (text.trim()) return text;
    } catch {
      continue;
    }
  }
  throw new Error("voice");
}

function clipVoice(raw: string): string {
  const prompt = raw.trim();
  if (!prompt || [...prompt].length > MAX_EXTRA) throw new Error("voice");
  return prompt;
}

File: src/llm/custom_voice.rs
//! Build a custom refine voice from a short operator interview.

use super::client::chat;
use super::endpoint::LlmEndpoint;
use super::refine::MAX_EXTRA_CHARS;
use super::refine_prompt::language_lock;
use super::types::{ChatMessage, ChatRequest, LlmError};
use serde::{Deserialize, Serialize};

const SYSTEM: &str = "You write a Klar NLU voice block only. \
Output the voice description and a few short example rewrites (source → spoken). \
Keep language lock and safety: no Home Assistant tools, no device control, \
digits stay digits, no new facts. Do not repeat the safety rules. \
Do not mention this instruction. No markdown fences. No title.";

#[derive(Debug, Clone, Deserialize)]
pub struct CustomVoiceRequest {
    #[serde(default)]
    pub language: String,
    #[serde(default)]
    pub address: String,
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub tone: String,
    #[serde(default)]
    pub humor: String,
    #[serde(default)]
    pub length: String,
    #[serde(default)]
    pub taboo: String,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct CustomVoiceOut {
    pub prompt: String,
}

#[derive(Debug, Clone)]
pub struct SanitizedInterview {
    pub language: String,
    pub address: &'static str,
    pub name: String,
    pub tone: &'static str,
    pub humor: &'static str,
    pub length: &'static str,
    pub taboo: String,
}

impl CustomVoiceRequest {
    pub fn sanitize(self) -> Result<SanitizedInterview, LlmError> {
        let language = self.language.trim();
        if language.is_empty() || language.chars().count() > 32 || language.chars().any(char::is_control) {
            return Err(LlmError::InvalidRequest("language"));
        }
        let address = match self.address.trim() {
            "du" => "du",
            "sie" => "sie",
            "name" => "name",
            _ => return Err(LlmError::InvalidRequest("address")),
        };
        let name = self.name.trim();
        if name.chars().count() > 64 || name.chars().any(char::is_control) {
            return Err(LlmError::InvalidRequest("name"));
        }
        if address == "name" && name.is_empty() {
            return Err(LlmError::InvalidRequest("name"));
        }
        let tone = match self.tone.trim() {
            "short" => "short",
            "warm" => "warm",
            "dry" => "dry",
            _ => return Err(LlmError::InvalidRequest("tone")),
        };
        let humor = match self.humor.trim() {
            "none" => "none",
            "light" => "light",
            "sharp" => "sharp",
            _ => return Err(LlmError::InvalidRequest("humor")),
        };
        let length = match self.length.trim() {
            "one" => "one",
            "more" => "more",
            _ => return Err(LlmError::InvalidRequest("length")),
        };
        let taboo = self.taboo.trim();
        if taboo.chars().count() > 200 || taboo.chars().any(|ch| ch.is_control() && ch != '\n' && ch != '\t') {
            return Err(LlmError::InvalidRequest("taboo"));
        }
        Ok(SanitizedInterview {
            language: language.to_string(),
            address,
            name: name.to_string(),
            tone,
            humor,
            length,
            taboo: taboo.to_string(),
        })
    }
}

pub fn interview_user_line(body: &SanitizedInterview) -> String {
    let address = if body.address == "name" { format!("address by first name ({})", body.name) } else { body.address.to_string() };
    let mut line = format!(
        "Language pack: {}.\nAddress: {}.\nTone: {}.\nHumor: {}.\nLength: {}.",
        body.language, address, body.tone, body.humor, body.length
    );
    if !body.taboo.is_empty() {
        line.push_str("\nDo not say: ");
        line.push_str(&body.taboo);
    }
    line
}

pub fn clip_voice(raw: &str) -> Result<String, LlmError> {
    let prompt = raw.trim().to_string();
    if prompt.is_empty() || prompt.chars().count() > MAX_EXTRA_CHARS {
        return Err(LlmError::Response);
    }
    Ok(prompt)
}

pub async fn generate_custom_voice(endpoint: &LlmEndpoint, request: CustomVoiceRequest) -> Result<CustomVoiceOut, LlmError> {
    let body = request.sanitize()?;
    let lock = language_lock(&body.language);
    let chat_req = ChatRequest {
        messages: vec![ChatMessage::new("system", format!("{SYSTEM}\n\n{lock}")), ChatMessage::new("user", interview_user_line(&body))],
        stream: Some(false),
        temperature: Some(0.4),
        max_tokens: Some(512),
        tools: None,
        tool_choice: None,
    };
    let prompt = clip_voice(&chat(endpoint, chat_req).await?)?;
    Ok(CustomVoiceOut { prompt })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rejects_unknown_choices() {
        let bad = CustomVoiceRequest {
            language: "de".into(),
            address: "hey".into(),
            name: String::new(),
            tone: "warm".into(),
            humor: "none".into(),
            length: "one".into(),
            taboo: String::new(),
        };
        assert!(bad.sanitize().is_err());
    }

    #[test]
    fn formats_interview() {
        let body = CustomVoiceRequest {
            language: "de".into(),
            address: "name".into(),
            name: "Ines".into(),
            tone: "dry".into(),
            humor: "light".into(),
            length: "one".into(),
            taboo: "chef".into(),
        }
        .sanitize()
        .unwrap();
        let line = interview_user_line(&body);
        assert!(line.contains("Language pack: de."));
        assert!(line.contains("Ines"));
        assert!(line.contains("Do not say: chef"));
    }

    #[test]
    fn empty_generated_prompt_fails() {
        assert!(clip_voice("   ").is_err());
        assert!(clip_voice(&"x".repeat(MAX_EXTRA_CHARS + 1)).is_err());
        assert_eq!(clip_voice("Voice: dry.").unwrap(), "Voice: dry.");
    }
}
